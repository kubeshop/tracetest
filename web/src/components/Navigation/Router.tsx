import {BrowserRouter, Routes, Route, Navigate} from 'react-router-dom';
import Home from 'pages/Home';
import Test from 'pages/Test';
import Trace from '../../pages/Trace';

const Router = (): JSX.Element => {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Home path="home" />} />
        <Route path="/test/:testId" element={<Test path="test-details" />} />
        <Route path="/test/:testId/run/:runId" element={<Trace path="trace" />} />
        <Route path="*" element={<Navigate to="/" />} />
      </Routes>
    </BrowserRouter>
  );
};

export default Router;
