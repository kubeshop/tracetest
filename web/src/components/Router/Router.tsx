import {Navigate, Route, Routes} from 'react-router-dom';

import Layout from 'components/Layout/Layout';
import VariableSet from 'pages/VariableSet';
import Home from 'pages/Home';
import TestSuites from 'pages/TestSuites';
import RunDetail from 'pages/RunDetail';
import Settings from 'pages/Settings';
import Test from 'pages/Test';
import TestSuite from 'pages/TestSuite';
import TestSuiteRunOverview from 'pages/TestSuiteRunOverview';
import TestSuiteRunAutomate from 'pages/TestSuiteRunAutomate';
import AutomatedTestRun from 'pages/AutomatedTestRun';
import CreateTest from 'pages/CreateTest';
import Wizard from 'pages/Wizard';
import {useDashboard} from 'providers/Dashboard/Dashboard.provider';

const Router = () => {
  const {baseUrl} = useDashboard();

  return (
    <Routes>
      <Route element={<Layout hasMenu />}>
        <Route path="/" element={<Wizard />} />
        <Route path="/tests" element={<Home />} />
        <Route path="/testsuites" element={<TestSuites />} />
        <Route path="/variablesets" element={<VariableSet />} />
        <Route path="/settings" element={<Settings />} />
      </Route>

      <Route element={<Layout />}>
        <Route path="/test/create/:triggerType" element={<CreateTest />} />
        <Route path="/test/:testId" element={<Test />} />
        <Route path="/test/:testId/run/:runId" element={<RunDetail />} />
        <Route path="/test/:testId/run/:runId/:mode" element={<RunDetail />} />
        <Route path="/test/:testId/run" element={<AutomatedTestRun />} />

        <Route path="/testsuite/:testSuiteId" element={<TestSuite />} />
        <Route path="/testsuite/:testSuiteId/run/:runId" element={<TestSuiteRunOverview />} />
        <Route path="/testsuite/:testSuiteId/run/:runId/overview" element={<TestSuiteRunOverview />} />
        <Route path="/testsuite/:testSuiteId/run/:runId/automate" element={<TestSuiteRunAutomate />} />
      </Route>

      <Route path="*" element={<Navigate to={baseUrl || '/'} />} />
    </Routes>
  );
};

export default Router;
