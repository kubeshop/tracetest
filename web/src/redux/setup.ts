import {Middleware} from '@reduxjs/toolkit';

import OtelRepoAPI from 'redux/apis/OtelRepo';
import TestSpecs from 'redux/slices/TestSpecs.slice';
import Spans from 'redux/slices/Span.slice';
import DAG from 'redux/slices/DAG.slice';
import Trace from 'redux/slices/Trace.slice';
import CreateTestSuite from 'redux/slices/CreateTestSuite.slice';
import User from 'redux/slices/User.slice';
import TestOutputs from 'redux/testOutputs/slice';

export const middlewares: Middleware[] = [OtelRepoAPI.middleware];

export const reducers = {
  [OtelRepoAPI.reducerPath]: OtelRepoAPI.reducer,

  spans: Spans,
  dag: DAG,
  trace: Trace,
  testSpecs: TestSpecs,
  createTestSuite: CreateTestSuite,
  user: User,
  testOutputs: TestOutputs,
};
