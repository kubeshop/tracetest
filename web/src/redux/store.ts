import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import TestAPI from 'redux/apis/Test.api';
import ResultListSlice from 'redux/slices/ResultList.slice';

export const store = configureStore({
  reducer: {
    [TestAPI.reducerPath]: TestAPI.reducer,
    resultList: ResultListSlice,
  },
  middleware: getDefaultMiddleware => getDefaultMiddleware().concat(TestAPI.middleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
