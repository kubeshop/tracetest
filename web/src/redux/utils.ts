import {PromiseWithKnownReason} from '@reduxjs/toolkit/dist/query/core/buildMiddleware/types';
import webSocketGateway, {IListenerFunction} from 'gateways/WebSocket.gateway';

export type {IListenerFunction} from 'gateways/WebSocket.gateway';

interface IInitWebSocketSubscription {
  listener: IListenerFunction;
  resource: string;
  waitToCleanSubscription: Promise<void>;
  waitToInitSubscription: PromiseWithKnownReason<any, any>;
}

/**
 * Implements RTK Query logic to receive streaming updates for persistent queries.
 * It allows a query to establish a WebSocket subscription, and apply updates to
 * the cached data when the information is received from the server.
 * For more information please take a look: https://redux-toolkit.js.org/rtk-query/usage/streaming-updates
 */
export async function initWebSocketSubscription({
  listener,
  resource,
  waitToCleanSubscription,
  waitToInitSubscription,
}: IInitWebSocketSubscription) {
  try {
    await waitToInitSubscription;
    webSocketGateway.subscribe(resource, listener);
  } catch {
    // no-op in case `waitToCleanSubscription` resolves before `waitToInitSubscription`,
    // in which case `waitToInitSubscription` will throw
  }
  await waitToCleanSubscription;
  webSocketGateway.unsubscribe(resource, '1234');
}
