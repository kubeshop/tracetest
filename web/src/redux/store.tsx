import {Action, configureStore, ThunkAction} from '@reduxjs/toolkit';
import {createReduxHistoryContext} from 'redux-first-history';
import {createBrowserHistory} from 'history';
import TracetestAPI from 'redux/apis/Tracetest';

import RouterMiddleware from './Router.middleware';
import {middlewares, reducers} from './setup';

const {createReduxHistory, routerMiddleware, routerReducer} = createReduxHistoryContext({
  history: createBrowserHistory(),
});

TracetestAPI.create();

export const store = configureStore({
  reducer: {
    ...reducers,
    [TracetestAPI.instance.reducerPath]: TracetestAPI.instance.reducer,
    router: routerReducer,
  },
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware()
      .prepend(RouterMiddleware.middleware)
      .concat(TracetestAPI.instance.middleware, ...middlewares, routerMiddleware),
});

export const history = createReduxHistory(store);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
