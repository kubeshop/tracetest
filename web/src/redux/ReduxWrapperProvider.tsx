import * as React from 'react';
import {Provider} from 'react-redux';
import {store} from './store';

interface IProps {
  children: React.ReactNode;
}

export const ReduxWrapperProvider = ({children}: IProps) => {
  return <Provider store={store}>{children}</Provider>;
};
