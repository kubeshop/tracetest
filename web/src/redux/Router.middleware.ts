import {createListenerMiddleware} from '@reduxjs/toolkit';
import type {TypedStartListening} from '@reduxjs/toolkit';
import {LOCATION_CHANGE} from 'redux-first-history';
import {parse} from 'query-string';
import RouterActions from './actions/Router.actions';
import {runOriginPathAdded} from './slices/User.slice';
import {AppDispatch, RootState} from './store';

type AppStartListening = TypedStartListening<RootState, AppDispatch>;
const listener = createListenerMiddleware();
const startAppListening = listener.startListening as AppStartListening;

const sideEffectActionList = [RouterActions.updateSelectedAssertion, RouterActions.updateSelectedSpan];

const runUrlRegex = /^\/(test|testsuite)\/([^\/]+)\/run\/([^\/]+)(.*)$/;

const RouterMiddleware = () => ({
  middleware: listener.middleware,

  startListening(params = {}) {
    return startAppListening({
      predicate: ({type = ''}) =>
        type === LOCATION_CHANGE || (type.endsWith('/fulfilled') && !type.startsWith('router/')),
      effect(_, {dispatch, getState}) {
        const {
          router: {location},
        } = getState();

        const search = parse(location?.search || '');

        sideEffectActionList.forEach(action => {
          dispatch(action({search, params}));
        });
      },
    });
  },

  startListeningForLocationChange() {
    return startAppListening({
      predicate: action => {
        const pathname = action?.payload?.location?.pathname ?? '';
        return action?.type === LOCATION_CHANGE && pathname.match(runUrlRegex);
      },
      effect: async (_, {dispatch, getOriginalState, getState}) => {
        const {
          router: {location: prevLocation},
        } = getOriginalState();

        const {
          router: {location: currLocation},
        } = getState();

        const prevPathname = prevLocation?.pathname ?? '';

        if (!prevPathname.match(runUrlRegex) && !prevPathname.includes('/create')) {
          const defaultPath = currLocation?.pathname?.includes('testsuite') ? '/testsuites' : '/tests';
          dispatch(runOriginPathAdded(prevLocation?.pathname ?? defaultPath));
        }
      },
    });
  },
});

export default RouterMiddleware();
