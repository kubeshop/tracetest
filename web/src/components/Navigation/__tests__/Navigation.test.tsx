import {render, waitFor} from '@testing-library/react';
import {ReduxWrapperProvider} from '../../../redux/ReduxWrapperProvider';
import Router from '../Router';

jest.mock('../../../services/Analytics/Analytics.service', () => {
  return {
    event: jest.fn(),
  };
});

test('Router', async () => {
  const {getByText, getByTestId} = render(
    <ReduxWrapperProvider>
      <Router />
    </ReduxWrapperProvider>
  );

  await waitFor(() => getByTestId('github-link'));
  expect(getByText('Documentation')).toBeTruthy();
});
