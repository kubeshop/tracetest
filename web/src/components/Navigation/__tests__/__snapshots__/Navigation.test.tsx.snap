// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router 1`] = `
<div>
  <header
    class="ant-layout-header sc-gsDKAQ fqXdzP"
  >
    <a
      href="/"
    >
      <h2
        class="ant-typography sc-dkPtRN jStClu"
      >
        <img
          alt="tracetest_log"
          data-cy="logo"
          src="Logo.svg"
        />
      </h2>
    </a>
    <ul
      class="ant-menu ant-menu-root ant-menu-horizontal ant-menu-light sc-hKwDye gtboAf"
      data-menu-list="true"
      role="menu"
      tabindex="0"
    >
      <li
        class="ant-menu-item ant-menu-item-only-child"
        data-menu-id="rc-menu-uuid-test-github"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="ant-menu-title-content"
        >
          <a
            data-cy="github-link"
            href="https://github.com/kubeshop/tracetest"
            target="_blank"
          >
            GitHub
          </a>
        </span>
      </li>
      <li
        class="ant-menu-item ant-menu-item-only-child"
        data-menu-id="rc-menu-uuid-test-docs"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="ant-menu-title-content"
        >
          <a
            data-cy="documentation-link"
            href="https://kubeshop.github.io/tracetest/"
            target="_blank"
          >
            Documentation
          </a>
        </span>
      </li>
      <li
        class="ant-menu-submenu ant-menu-submenu-horizontal"
        role="none"
      >
        <div
          aria-controls="rc-menu-uuid-test-SubMenu-popup"
          aria-expanded="false"
          aria-haspopup="true"
          class="ant-menu-submenu-title"
          data-menu-id="rc-menu-uuid-test-SubMenu"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-menu-title-content"
          >
            <span
              aria-label="question-circle"
              class="anticon anticon-question-circle"
              data-cy="onboarding-link"
              role="img"
              style="color: rgb(97, 23, 94); font-size: 16px;"
            >
              <svg
                aria-hidden="true"
                data-icon="question-circle"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                />
                <path
                  d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z"
                />
              </svg>
            </span>
          </span>
          <i
            class="ant-menu-submenu-arrow"
          />
        </div>
      </li>
    </ul>
    <div
      aria-hidden="true"
      style="display: none;"
    />
  </header>
  <main
    class="ant-layout-content sc-bdvvtL cVXBSg"
  >
    <div
      class="sc-llYSUQ jPRqiW"
    >
      <h4
        class="ant-typography sc-cxpSdN bbcJQg"
      >
        All Tests
      </h4>
      <div
        class="sc-iwjdpV NicDQ"
      >
        <span
          class="ant-input-affix-wrapper sc-gWXbKe kPVVws"
        >
          <span
            class="ant-input-prefix"
          >
            <span
              aria-label="search"
              class="anticon anticon-search"
              role="img"
              style="color: rgba(3, 24, 73, 0.4);"
            >
              <svg
                aria-hidden="true"
                data-icon="search"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                />
              </svg>
            </span>
          </span>
          <input
            class="ant-input"
            height="32px"
            placeholder="Search test (Not implemented yet)"
            type="text"
            value=""
            width="270px"
          />
        </span>
        <div
          class="sc-iJKOTD bLqiN"
        >
          <button
            class="ant-btn ant-btn-primary sc-bBHxTw"
            data-cy="create-test-button"
            data-tour="home_create_test"
            type="button"
          >
            <span>
              Create Test
            </span>
          </button>
        </div>
      </div>
      <div
        class="sc-kLwhqv kFBBsa"
        data-cy="test-list"
      />
    </div>
    <div
      class="sc-ikJyIC exAstU"
    />
  </main>
</div>
`;
